compliance/collection/nodeinventorizer/caching_inventorizer.go

```
@@ -0,0 +1,284 @@
+	Inventory       *storage.NodeInventory
+}
+
+// NewCachingScanner returns a ready to use instance of Caching Scanner
+func NewCachingScanner(inventoryCachePath string, cacheDuration time.Duration, initialBackoff time.Duration, maxBackoff time.Duration, backoffCallback func(time.Duration)) *CachingScanner {
+	return &CachingScanner{
+		inventoryCachePath:  inventoryCachePath,
+		cacheDuration:       cacheDuration,
+		initialBackoff:      initialBackoff,
+		maxBackoff:          maxBackoff,
+		backoffWaitCallback: backoffCallback,
+	}
+}
+
+// Scan scans the current node and returns the results as storage.NodeInventory struct.
+// A cached version is returned if it exists and is fresh enough.
+// Otherwise, a new scan guarded by a backoff is run.
+func (c *CachingScanner) Scan(nodeName string) (*storage.NodeInventory, error) {
+	// check whether a cached inventory exists that has not exceeded its validity
+	cache := readInventoryWrap(c.inventoryCachePath)
+	if cache != nil && cache.Inventory != nil && cache.ValidUntil.After(time.Now()) {
+		log.Debugf("Using cached node scan (valid until %v)", cache.ValidUntil)
+		return cache.Inventory, nil
+	}
+
+	// check for existing backoff, wait for specified duration if needed, then persist the new backoff duration
+	backoffDuration := c.initialBackoff
+	if cache != nil {
+		backoffDuration = c.validateBackoff(cache.BackoffDuration)
+	}
+
+	if backoffDuration > c.initialBackoff {
+		log.Warnf("Found existing node scan backoff file - last scan may have failed. Waiting %v seconds before retrying", backoffDuration.Seconds())
+		c.backoffWaitCallback(backoffDuration)
+	}
+
+	// Write backoff duration to cache
+	backoff := inventoryWrap{BackoffDuration: c.calcNextBackoff(backoffDuration)}
+	if err := writeInventoryWrap(backoff, c.inventoryCachePath); err != nil {
+		log.Warnf("Error writing node scan backoff file: %v", err)
+	}
```

Sorry for back-and-forth on this logic and commenting on something that was already there for some time. Although, the current code is becoming more clear than former iterations, and things like the following become more evident due to this.

Consider a scenario when our code can't write to cache directory at all (either due to no mount present or due to a full volume, etc.) In this case, error writing backoff is swallowed (with warning) and the code proceeds to do the real scan i.e. loads the node. Assuming scanning succeeds, next the code will try to save the cache which fails and returns an error.

What happens next? On retry, the node is scanned once again without applying a backoff.

I think, an error here should lead to propagation rather than a warning to fix the described scenario:

```suggestion
	if err := writeInventoryWrap(backoff, c.inventoryCachePath); err != nil {
		return nil, errors.Wrap(err, " writing node scan backoff file")
	}
```

The observed end result will be the same - no scan information, but at least this will not load the node in case of retries.